### DISCUSSION TITLE
Thread safety issue while using basicConsume in lambdas

### DISCUSSION THREAD

**Discussion by brothergod1** [rabbitmq/rabbitmq-java-client#979] [General]
I have a KeyCloak application which as you may know uses Hibernate for entity management. The reason why I mention it is that in DeliverCallback function the client tries to save user with the data provided in this notification, but instead of that it fails an throws an exception "IllegalStateException: Session / Entity Manager is closed". The exception is thrown only in case when the code that is responsible for saving a user is inside the DeliverCallback function and knowing that I might say that the thread safety here is doubtful. I also tries the way of polling the notifications, but it doesn't work properly. Is there any other way to subscribe for notifications except the polling and using the functional interface?

**Comment by michaelklishin** [comment]
The client does not synchronize its calls to `handleDelivery`. You haven't provided anything beyond a speculation but if your consumer handler code needs synchronisation, that is on you as application developer to do in whatever manner would be appropriate.

**Reply by acogoluegnes** [reply]
How is the entity manager usually open and closed in your application? Whatever the mechanism, it is not called automatically in the delivery callback. Common mechanisms are a servlet filter in a web application or an AOP interceptor in a Spring application. You may have to handle the entity manager lifecycle manually for incoming messages.
